forward_authentication_service/PreAuth/demo-preauth.sh

```diff
@@ -150,26 +165,49 @@ fi
 #
 if [ -z $username ] || [ -z $emailaddr ]; then
 	echo "<big-red>Welcome!</big-red><italic-black> To access the Internet you must enter your Name and Email Address</italic-black><hr>"
-	echo "<form action=\"/nodogsplash_preauth/\" method=\"get\">"
-	echo "<input type=\"hidden\" name=\"clientip\" value=\"$clientip\">"
-	echo "<input type=\"hidden\" name=\"gatewayname\" value=\"$gatewayname\">"
-	echo "<input type=\"hidden\" name=\"redir\" value=\"$requested\">"
-	echo "<input type=\"text\" name=\"username\" value=\"$username\" autocomplete=\"on\" >:Name<br><br>"
-	echo "<input type=\"email\" name=\"emailaddr\" value=\"$emailaddr\" autocomplete=\"on\" >:Email<br><br>"
-	echo "<input type=\"submit\" value=\"Continue\" >"
-	echo "</form><hr>"
+	echo -e $login_form
 else
-	# We have both fields, so get the token and in this case also the mac address
-	tok="$(ndsctl json $clientip | grep token | cut -c 10- | cut -c -8)"
-	clientmac="$(ndsctl json $clientip | grep mac | cut -c 8- | cut -c -17)"
+	# If we got here, we have both the username and emailaddr fields as completed on the login page on the client,
+	# so we will now call ndsctl to get client data we need to authenticate and add to our log.
+
+	# Variables returned from ndsctl are listed in $varlist.
+
+	# We at least need the client token to authenticate.
+	# In this example we will also log the client mac address.
+
+	varlist="id ip mac added active duration token state downloaded avg_down_speed uploaded avg_up_speed"
```

Comment:
@mwarning 
json parser

Reply:
Sorry - That didn't work - How do you get github to display a specific bit of code?

Reply:
This is the json parser with check for ndsctl empty response in case of another instance of nds already running.

```
	varlist="id ip mac added active duration token state downloaded avg_down_speed uploaded avg_up_speed"
	clientinfo=$(ndsctl json $clientip)

	if [ -z $clientinfo ]; then
		echo "<big-red>Sorry!</big-red><italic-black> The portal is busy, please try again.</italic-black><hr>"
		echo -e $login_form
		echo -e $footer
		exit 0
	else
		for var in $varlist; do
			eval $var=$(echo "$clientinfo" | grep $var | awk -F'"' '{print $4}')
		done
	fi

```